### DISCUSSION TITLE
How to use RichTextBox in WPF?

### DISCUSSION THREAD

**Discussion by VahidEra** [dotnet/wpf#7596] [Q&A] [answered]
Hi guys,
Rich Textbox in WPF is a completely different control from Rich Textbox in WinForm, and it has become a problem for me to work with it. I was running a code in winform that I do not know how to execute the same code in WPF because in winform I can easily count the number of lines and even the color of the specified length, I will write the code below and if possible I will run this code in WPF.

```
Private Sub rtb_MouseDown(sender As Object, e As MouseEventArgs) Handles rtb.MouseDown
        If e.Button = MouseButtons.Left Then
            rtb.Text = rtb.Text.Insert(rtb.SelectionStart, Environment.NewLine)
            Line0Count.Text = rtb.Lines(0).Length

            If rtb.Lines.Count > 1 Then
                Dim Line0Count As String = rtb.Lines(0).Length
                Dim Line1Count As String = rtb.Lines(1).Length
                If Line0Count > 38 Then
                    rtb.SelectionStart = 0
                    rtb.SelectionLength = Line0Count
                    rtb.SelectionColor = System.Drawing.Color.Red
                    rtb.SelectionStart = 0
                    rtb.SelectionLength = 0
                Else
                    rtb.SelectionStart = 0
                    rtb.SelectionLength = Line0Count
                    rtb.SelectionColor = System.Drawing.Color.Black
                    rtb.SelectionStart = 0
                    rtb.SelectionLength = 0
                End If
                If Line1Count > 38 Then
                    rtb.SelectionStart = Line0Count + 1
                    rtb.SelectionLength = Line1Count
                    rtb.SelectionColor = System.Drawing.Color.Red
                    rtb.SelectionStart = 0
                    rtb.SelectionLength = 0
                Else
                    rtb.SelectionStart = Line0Count + 1
                    rtb.SelectionLength = Line1Count
                    rtb.SelectionColor = System.Drawing.Color.Black
                    rtb.SelectionStart = 0
                    rtb.SelectionLength = 0
                End If
            End If
        End If
End sub
```

>thank you so much for your help.

**Comment by miloush** [comment, marked as answer]
To change a color of a range, for example selection, you can use ApplyPropertyValue:
```C#
_rtb.Selection.ApplyPropertyValue(TextBlock.ForegroundProperty, Brushes.Red);
```

To get a text of RTB, you need to create a text range first and then you can get or set text:
```C#
string text = new TextRange(_rtb.Document.ContentStart, _rtb.Document.ContentEnd).Text;
```

E.g. to insert a new line, at the selection start, you create a range and set text in it:
```C#
TextPointer selectionStart = _rtb.Selection.Start;
new TextRange(selectionStart, selectionStart).Text = Environment.NewLine;
```

to count lines, you either have to count lines in the text retrieved as above (e.g. splitting the string by new line into an array), or you would have to loop through `_rtb.Document.Blocks`. Depends how you count lines.

This should get you started.

**Reply by miloush** [reply]
Note that `RichTextBox.Document` is a `FlowDoucment`. You might find https://learn.microsoft.com/en-us/dotnet/desktop/wpf/advanced/flow-document-overview and surrounding pages helpful.

**Reply by VahidEra** [reply]
@miloush Dear Jan, As always, you gave me a complete, concise and flawless answer. I read many threads about WPF, but none of them gave me the right answer.  You completely understand what I mean and give me the correct answer. I almost learned WPF from you. Thank you very much, that was exactly what I wanted.
